safe_api.js

```diff
@@ -78,10 +133,41 @@ window.safeAPI = {};
 					.then(() => mdHandle));
 		}
 
+		getString(key) {
+			return this.then(_ => {
+				console.log('Fetching key from handle: ', key, this.mdHandle);
+				return safeMutableData.get(this.mdHandle, key);
+			})
+			.then((val) => {
+				return val.buf.toString();
+			});
+		}
+
 		free() {
-			this.handlePromise.then((mdHandle) => {
-				safeMutableData.free(mdHandle);
-				this.handlePromise = null;
+			this.then((mdHandle) => {
+				// Nothing to do if no handles are provided
+				if ((mdHandle === undefined) && (this.mdHandle === undefined)) {
+					console.log('No MD to free.');
+				};
+				// only this.mdHandle is defined, clear it.
+				if ((mdHandle === undefined) && (this.mdHandle != undefined)) {
+					safeMutableData.free(this.mdHandle);
+					console.log('Freed MD: ', this.mdHandle);
+					this.mdHandle = undefined;
+				};
+				// only mdHandle is defined, clear it.
+				if ((mdHandle != undefined) && (this.mdHandle === undefined)) {
+					safeMutableData.free(mdHandle);
```

Comment:
mdHandle will be appHandle in fact?

Reply:
`this.mdHandle` will be the `mdHandle`, no question.

You put `mdHandle` in as a function call, which I left. For any code you wrote, you're assuming that the previous all before `free()` is returning the `mdHandle`. It might be the `mdHandle`, it might be the `appHandle`, it entirely depends on the return of the previous `thenable` called.

This is why I started storing `self.handle`: no confusion.